### ISSUE TITLE
NKAI very stupid behavior when limited to 2 heroes

### ISSUE THREAD

**Issue by val-gaav** [vcmi/vcmi#3542]
**Describe the bug**
There seems to be a regression with NKAI when it is limited to 2 heroes. It worked ok maybe around vcmi 1.2 or 1.3 I think, but right now on Jebus cross AI just stays in its zone and only does some chaotic pointless movements. It doesn't meter what bonuses it has (extra resources etc.) Sometimes a hero is stuck going out of castle and then returning to it, and does the same action for several turns (save file included). Sometimes AI hero is stuck in place and it takes extra 5-10 seconds for AI to finish its turn.   Around month 4 it still moves around and the stuff in zone like pandora boxes , arts etc. are still uncleared (save file included). Heroes are low level even though they could do some serious fighting as they have the armies to do so.  I would say the level of stupidity is alike to VCAI or worse 😛
Probably this sometimes happen with 8 hero limit too. Just is a bit harder to trigger. For example I got a hero stuck in town on titans winter map.

Also as included saves show in some towns AI does not buy troops even though it has a lot of resources, and extra income. It gets 4000 gold extra daily from global bonuses. 

**To Reproduce**
Enable limit hero submod  from Boost AI , or limit heroes to two from gameconfig json file, start random jebus cross map and watch how AI plays. I've also included save files to show how AI playes after month and 4 months. Included is also a save called hero stuck . After loading it for several turns hero of AI player should be stuck in loop entering town 

**Expected behavior**
AI actually doing some goals and not getting stuck . It should pick up stuff and try to level up heroes it has. At some point it should try to break into center zone and attack other players.

**Actual behavior**
AI seems to get stuck on some tasks. It does not have clear goals as it used to and just wonders around in zone aimlessly. 

**Did it work earlier?**
Yes it used to work around year ago. Maybe version 1.2 or 1.3 ? Hard to tell when exactly as I was inactive for a while as far as vcmi gaming goes.  Generally AI was aggressive back then, was leveling up heroes and at some point it tried to get into central zone and even attack other players. 

**Version**
Linux Ubuntu deb package from vcmi ppa
 - Version: 1.4.4

**Comment by val-gaav** [comment]
[saves_.zip](https://github.com/vcmi/vcmi/files/14002872/saves_.zip)
[VCMI_logs.zip](https://github.com/vcmi/vcmi/files/14002873/VCMI_logs.zip)

**Comment by val-gaav** [comment]
I think I had one game with vcmi to check Factory town port , so it might be that in vcmi 1.4.0 nkai was still ok with 2 hero limit.

**Comment by val-gaav** [comment]
It seems AI plays normally in 1.4.2 so regression is either due to 1.4.3 or 1.4.4

**Comment by ninjatao** [comment]
> It seems AI plays normally in 1.4.2 so regression is either due to 1.4.3 or 1.4.4

In 1.4.4, in some map templates like diamond, AI will never conquer a second town. 1.4.2 AI works fine.

**Comment by DjWarmonger** [comment]
Yes, I observed yesterday that NKAI does not play agressively in develop and barely attacks anything. Also it rarely builds chains. Using my usual test map (Jebus), no AI mods, 8 heroes.

**Comment by DjWarmonger** [comment]
Maybe here? [lib/pathfinder/PathfindingRules.cpp](https://github.com/vcmi/vcmi/blob/0c1664dbe736749bb524b21ed954e805c81d6b87/lib/pathfinder/PathfindingRules.cpp#L251)

Or here? https://github.com/vcmi/vcmi/commit/708e0d11072f0ef88882e9df69314b96354b2cc9

**Comment by DjWarmonger** [comment]
Okay found it: https://github.com/vcmi/vcmi/commit/8303ce5d1321566abf8c39922de34d5836c04670
